### PR TITLE
Captcha api protection

### PR DESCRIPTION
**Pull request by samerbuna**
Converting #366 to a local repo branch

### CODE REVIEW

**Comment by samerbuna** [comment]
> Shoot, I just realized that because I integrated the auth code into the last release of the app that we can't make this breaking change to this endpoint anymore. Will need a different endpoint for requesting an authcode with geetest

Renamed

**Review by jrbuhl93 — APPROVED** [review]
Make sure we coordinate with @krtk6160 to get the environment variables for GeeTest in. We can used the ones I was using for testing for now, but when I upgrade to a paid plan on GeeTest they will need to be changed.

**Comment by daviroo** [comment]
Hey @samerbuna. I think this change requires one more `yarn install` commit.  When I checked it out locally to work on the mobile app changes it registers a change to the `yarn.lock` file.  Not a massive issue but it will help with idempotency.

**Comment by samerbuna** [comment]
> Hey @samerbuna. I think this change requires one more `yarn install` commit. When I checked it out locally to work on the mobile app changes it registers a change to the `yarn.lock` file. Not a massive issue but it will help with idempotency.

Good catch. Updated.

**Review by jrbuhl93 — APPROVED** [review]

File: src/graphql/main/schema.graphql [hunk 1]
```diff
@@ -346,6 +346,8 @@ type Mutation {
   ): LnInvoicePaymentSendPayload!
   onChainAddressCreate(input: OnChainAddressCreateInput!): OnChainAddressPayload!
   onChainAddressCurrent(input: OnChainAddressCurrentInput!): OnChainAddressPayload!
+  captchaCreateChallenge: CaptchaCreteChallengePayload!
```
Comment (krtk6160) [inline src/graphql/main/schema.graphql:349]: Typo. Should be `CaptchaCreateChallengePayload`

**Review by krtk6160 — DISMISSED** [review]

File: src/graphql/main/schema.graphql (hunk 1 above)
Comment (samerbuna) [reply src/graphql/main/schema.graphql]: Good catch. Fixed.

**Comment by samerbuna** [comment]
Rebased to resolve conflicts: @jrbuhl93 @krtk6160

File: src/graphql/root/mutation/captcha-request-auth-code.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  args: {
+    input: { type: GT.NonNull(CaptchaRequestAuthCodeInput) },
+  },
+  resolve: async (_, args, { logger, ip, geetest }) => {
+    const {
+      phone,
+      challengeCode: geetestChallenge,
+      validationCode: geetestValidate,
+      secCode: geetestSeccode,
+    } = args.input
```
Comment (dolcalmi) [inline src/graphql/root/mutation/captcha-request-auth-code.ts:28]: does this need validation?
```js
for (const input of [...]) {
      if (input instanceof Error) {
        return { errors: [{ message: input.message }] }
      }
    }
```

File: src/graphql/root/mutation/captcha-request-auth-code.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    try {
+      // TODO: redo this with use-case and errors pattern
+      await verifyCaptchaAndReturnOTP({
```
Comment (dolcalmi) [inline src/graphql/root/mutation/captcha-request-auth-code.ts:32]: not all fails throw an exception, if `geetest.validate` fail it returns false

**Review by dolcalmi — CHANGES_REQUESTED** [review]
I could only test `captchaCreateChallenge` because mobile implementation has not been finished.

File: src/graphql/root/mutation/captcha-request-auth-code.ts (hunk 3 above)
Comment (samerbuna) [reply src/graphql/root/mutation/captcha-request-auth-code.ts:32]: There is an if statement in `verifyCaptchaAndReturnOTP` for that case

Comment (dolcalmi) [reply src/graphql/root/mutation/captcha-request-auth-code.ts:38]: yes, that is the point, it returns false (not an exception) and in that case we are returning `success: true`

**Review by dolcalmi — APPROVED** [review]

**Review by jrbuhl93 — APPROVED** [review]

**Pull request merged**